src/base.js

```diff
@@ -323,7 +323,84 @@ class Base {
   /**
    * Returns a promise
    */
-  handleThirdPartyRoomMessage(thirdPartyRoomMessageData, doNotTryToGetRemoteUserStoreData) {
+  getUserClient(roomId, senderId, senderName, avatarUrl, text, doNotTryToGetRemoteUserStoreData) {
+    const { info } = debug(this.getUserClient.name);
+    info("get user intent for third party user %s (%s)", senderId, senderName);
+
+    if (senderId === undefined) {
+      return Promise.resolve(this.puppet.getClient());
+    } else {
+      if (!senderName && !this.allowNullSenderName) {
+        if (doNotTryToGetRemoteUserStoreData)
+          throw new Error('preventing an endless loop');
+
+        info("no senderName provided with payload, will check store");
+        return this.getOrInitRemoteUserStoreDataFromThirdPartyUserId(senderId).then((remoteUser)=>{
+          info("got remote user from store, with a possible client API call in there somewhere", remoteUser);
+          info("will retry now");
+          const senderName = remoteUser.get('senderName');
+          return this.getUserClient(roomId, senderId, senderName, true);
+        });
+      }
+
+      info("this message was not sent by me");
+      const ghostIntent = this.getIntentFromThirdPartySenderId(senderId);
+      let promiseList = [];
+      promiseList.push(() => ghostIntent.join(roomId));
+
+      if (senderName)
+        promiseList.push(() => ghostIntent.setDisplayName(senderName));
+
+      if (avatarUrl)
+        promiseList.push(() => this.setGhostAvatar(ghostIntent, avatarUrl));
+
+      return Promise.all(promiseList).then(() => ghostIntent.getClient());
```

Comment:
I will explain `p=>p()`. So mapSeries takes 2 args. The first argument is any list, and the 2nd argument is what to do with that list. The list does NOT have to be a list of promises. The 2nd argument MUST be a function returning a promise. What mapSeries does is it queues up the iteration through the list until the function of the 2nd argument resolves or rejects. In this way, everything is done in sequence. So in our case `p=>p()` we are actually iterating through a list of promises, so our 2nd argument is nothing but to execute the promise, hence `p()`

* `mapSeries` in bluebird's docs http://bluebirdjs.com/docs/api/promise.mapseries.html